flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/HeapStateStore.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Java heap backed {@link StateStore}.
+ *
+ * @param <T> Type of state
+ */
+class HeapStateStore<T extends Serializable> implements StateStore<T> {
+
+	private final Map<String, T> stateMap = new HashMap<>();
+
+	private final AtomicInteger idCounter = new AtomicInteger();
+
+	@Override
+	public String putState(T state) throws Exception {
+		checkNotNull(state, "State");
+
+		String key = "jobmanager://savepoints/" + idCounter.incrementAndGet();
```

Comment:
Why do we use an `AtomicInteger` here? Does this mean that `putState` will be accessed concurrently? If this is the case, then `stateMap.put` is problematic, since a `HashMap` is not thread safe.